contributors/design-proposals/auth/external-secrets.md

```diff
@@ -0,0 +1,218 @@
+authenticate to the external vault / external secrets provider.
+
+## User Experience
+
+### Use Cases
+
+* Access to TLS certificate for the Web application running in the
+  Pod / container; the certificate can be either in FreeIPA Vault
+  or generated or renewed on-the-fly via FreeIPA CA; Custodia is used
+  as the API.
+
+Let's assume a FUSE filesystem which provides access to its individual
+files as requests for certificates is implemented as `/usr/bin/secrets-fuse`.
+
+With (trivial for clarity) flexVolume plugin
+
+```
+#!/bin/bash
+
+# Put to
+# /usr/libexec/kubernetes/kubelet-plugins/volume/exec/example.com~secrets-fuse/secrets-fuse
+
+case "$1" in
+	init)
+		echo '{ "status": "Success", "capabilities": {"attach": false, "selinuxRelabel": false}}'
+	;;
+	mount)
+		mkdir -p "$2"
+		# TODO: parse kubernetes.io/pod.uid and pass it
+		mount -t fuse -o allow_other /usr/bin/secrets-fuse "$2"
+		echo '{ "status": "Success" }'
+	;;
+	unmount)
+		umount -f "$2"
+		rmdir "$2"
+		echo '{ "status": "Success" }'
+	;;
+	*)
+		echo '{ "status": "Not supported" }'
+		exit 1
+esac
+```
+
+the FUSE filesystem can get mounted to Pod like
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test-pod
+spec:
+  containers:
+  - image: registry.access.redhat.com/rhel7
+    imagePullPolicy: Never
+    name: test-pod
+    volumeMounts:
+    - mountPath: /mnt/secrets-external
+      name: secrets-external
+  volumes:
+  - flexVolume:
+      driver: example.com/secrets-fuse
+    name: secrets-external
+```
+
+Within the container of the pod, the application can then access
+`/mnt/secrets-external/HTTP/www.example.com` and the FUSE filesystem
+will see this as access to `HTTP/www.example.com` file. Via the
+`kubernetes.io/pod.uid`, it can verify (against its internal rules
+or for example via Kubernetes API against attributes of that Pod)
+whether the Pod should be allowed access to the `HTTP/www.example.com`
+secret, and then invoke tools like `custodia-cli` via
+
+```
+custodia-cli get HTTP/www.example.com
+```
+
+to get the secret from remote, external source. Alternatively, the
+Pod identity with other attributes / annotations can be passed to the
+remote provider to run the authorization on the remote side.
+
+Since the FUSE provider runs as root, it will have access to necessary
+credentials to prove its own identity to the remote provider.
+
+In production setup, we expect that remote provider to be a Custodia
+server which will make additional verification against Kubernetes API,
+checking that the on-Node identity which makes the request for
+secrets is only making the request on behalf of the Pods that are
+currently running on the Node. That way, if a Node is compromised,
+only (secrets of) Pods running on that single Node are compromised.
+The Custodia server on that separate host will then proxy the requests
+to the actual vault or secrets provider.
+
+## Implementation
+
+Implementation of FUSE filesystem using libfuse 2.9 will show example
```

Comment:
This is pretty low level given the proposal above - would expect you to mention that you have a particular implementation proposed, but the mechanism whereby you are implementing this allows other approaches (and list and describe them here, including drawbacks / advantages).

Reply:
Do you mean different versions of `libfuse` and different FUSE client modules for different languages?